onmt/translate/translator.py

```diff
@@ -180,6 +180,9 @@ def translate(
             image_channel_size=self.image_channel_size,
         )
 
+        assert not self.copy_attn or data.can_copy, \
+            "Copy attention is set, but dataset was trained without it"
+
```

Comment:
this was not asserted before. How was it behaving in this case ? (copy_attn set while not used in training)

Reply:
This is to make sure that if a model is trained with copy attention, the translator doesn't ignore that copy attention at translation time and translate as though it didn't exist. I don't _think_ this issue would actually occur in practice, but I would want to know if it did.

Reply:
Well, okay, I might have confused myself with the explanation I gave earlier. What is actually getting asserted is that if your model uses copy attention, the dataset must be compatible with copy attention (it needs to have the src_map field and stuff).

This probably still isn't necessary right now because the translation dataset is constructed using the fields from training time, but it is not difficult to imagine that not being the case (for example, allowing the translation dataset to be created ahead of time so it doesn't need to be tokenized each time you call translate.py).

I can remove the assertion if you want, but the issue has been on my mind because of the various issues around copy attention fields being missing when loading from old data.

Reply:
Well, IMO it makes the code harder to read.
I think it should be removed AND all below we should not test on this "can_copy" but keep the self.data_type == 'text'  since it is accessible (still defined in build_dataset)
no ?

Reply:
I'll remove the assert for now because the scenario in which it is useful does not exist yet. However, I don't see how it makes the code less readable.

At the current moment, I'm not sure either `self.data_type` or `self.data.can_copy` is necessary for the test. But if one were to be required, I would rather it be `can_copy` because it has an easy natural language interpretation and doesn't needed to be changed if we introduce other data types that allow copying, or change the usage of the `-data_type` command line argument. On the second point, I think `data_type` is kind of a misleading name because it doesn't refer to the data as a whole, only to the source. 

Besides all of which, it's not clear why a translator object should need to keep track of its `data_type`. So I'd prefer not to make references to it when I can avoid them.